### ISSUE TITLE
Revisit the 'mystery constant' in M6532.cxx

### ISSUE THREAD

**Issue by DirtyHairy** [stella-emu/stella#97]
According to commit history, the 'mystery constant' from Z26 was reintroduced to fix a hang in Mr. Roboto, while a simple comparision `timer >= 0` is all that should be necessary.

      if(!(timer & 0x40000))
      {
        // Return at 'divide by TIMxT' interval rate
        return (timer >> myIntervalShift) & 0xff;
      }

While working on fixing the remaining bugs in the 6502.ts RIOT implementation, I have made an interesting observation: while I promptly got the hang after the intro screen in 6502.ts (it is fixed now), I cannot reproduce it in Stella by changing the code to `timer >= 0` anymore!

Analysis suggests that RSYNC actually was the culprit and that our new RSYNC fixes the issues, while the old code above is more of a band aid to avoid the side effects :smirk: What happens is the following: after the title screen, Mr. Roboto writes to TIM64T, then goes through the canonical write-zero-to-everything-in-page-zero loop, supposedly from the original Bezerk code. In this init loop, RSYNC is anadvertedly hit and, if implemented correclty, fast-forwards the TIA playfield counter. Further down the lane, the code does a WSYNC, and the number of cycles that is skipped depends on the preceeding RSYNC. The hang is then caused by a loop that polls INTIM until it is zero. At this point, the counter will have wrapped and be counting down each cycle --- if timing is off, it will never read zero, and the game hangs :8ball:

The old implementation gets out of this situation by switching back to counting in TIM64T mode after bit 18 of `timer` flips again --- in this mode, the zero value is held for 64 cycles, which is long enough to escape the loop. With the new RSYNC simulation, timing works out in the right way, and the loop exits naturally.

@sa666666 , do you have any other examples of ROMs that failed with `timer > 0`? If my assumptions are correct, these should work now, too, and we can remove the cludge (which I strongly suspect differs from real hardware after bit 18 flips for the second time).

**Comment by sa666666** [comment]
I'm sorry, but I don't remember any of the other ROMs that required the old code.  If we want to test this, then I suggest making the change and having it part of pre5.  Then if we get regressions we can test the change on the reported ROMs.

In any event, your analysis makes sense to me.  I really like it when fixing a bug in one part of the system fixes a kludge somewhere else.  It means that the emulation is coming together, and each part of the system is depending on the other parts correctly.

**Comment by DirtyHairy** [comment]
:smile: OK, I have remove the kludge in 8f93704 --- let's see what comes of it...

**Comment by sa666666** [comment]
This constant was present in two other places too; they are fixed in https://github.com/stella-emu/stella/commit/2a2797cbafe28825755df5aeb1ccbc672673502b.

**Comment by sa666666** [comment]
Acid ROM fails because of this change, so I'm reopening this.  Though it may be an issue in the ROM, not the change itself.

[acid_drop.zip](https://github.com/stella-emu/stella/files/818830/acid_drop.zip)

**Comment by sa666666** [comment]
Another one that's broken (after pressing Fire/Space):

[panda_chase.zip](https://github.com/stella-emu/stella/files/818873/panda_chase.zip)

**Comment by DirtyHairy** [comment]
Very interesting. I can confirm that both run on real hardware, and at least Panda Chase (I took a closer look at that one) fails like Mr. Robot by going in and endless loop polling INTIM for zero while the counter wraps over and over again. I see two possiblities: either there is an aspect of RIOT behavior that we do not know about yet, or there is a timing glitch somewhere that causes the mismatch (one clock would suffice).

For future reference analysis, some observations on Panda Chase:

* The relevant timer write is a STA to TIM64T at address 0xF041
* There is no RSYNC or WSYNC after the final write to TIM64T (before the hang)
* I have attached a trace captured by hacking 6502.ts below. The first instruction is the final write to TIM64T, and the last instruction is the first INTIM of the infinite loop. The breakpoint at 0xF0A1 is the check on the fire button, so setting a breakpoint at $F0A5 allow to reproduce the same trace.

```
    $F041:   STA $0296
    $F044:   LDA $0282
    $F047:   LSR
    $F048:   BCS $2D ; -> $F077
    $F077:   LDX #$00
    $F079:   LSR
    $F07A:   BCS $1F ; -> $F09B
    $F09B:   STX $DA
    $F09D:   LDA $E9
    $F09F:   BMI $3E ; -> $F0DF
(B) $F0A1:   LDA $3C
    $F0A3:   BMI $1A ; -> $F0BF
    $F0A5:   LDA $EB
    $F0A7:   BPL -$5B ; -> $F04E
    $F04E:   DEC $EB
    $F050:   LDA $C6
    $F052:   STA $EF
    $F054:   LDX #$9C
    $F056:   LDA #$00
    $F058:   STA $4E,X
    $F05A:   DEX
    $F05B:   BNE -$05 ; -> $F058
   
[loop cut out]

    $F05D:   LDA $EF
    $F05F:   STA $C6
    $F061:   JSR $F069
    $F069:   CLC
    $F06A:   ADC #$01
    $F06C:   CMP #$0A
    $F06E:   BCC $04 ; -> $F074
    $F074:   STA $AA
    $F076:   RTS
    $F064:   JSR $F83B
    $F83B:   LDX #$24
    $F83D:   LDA #$78
    $F83F:   STA $92
    $F841:   LDA #$48
    $F843:   STA $93
    $F845:   RTS
    $F067:   BPL $63 ; -> $F0CC
    $F0CC:   LDA #$00
    $F0CE:   STA $19
    $F0D0:   STA $1A
    $F0D2:   JSR $F424
    $F424:   LDX #$00
    $F426:   LDY $C4
    $F428:   LDA $FC18,Y
    $F42B:   STA $9E,X
    $F42D:   INY
    $F42E:   INY
    $F42F:   INY
    $F430:   INY
    $F431:   LDA $FC18,Y
    $F434:   STA $98,X
    $F436:   INY
    $F437:   INY
    $F438:   INY
    $F439:   INY
    $F43A:   LDA $FC18,Y
    $F43D:   STA $AC,X
    $F43F:   INX
    $F440:   CPX #$02
    $F442:   BEQ $08 ; -> $F44C
    $F444:   CLC
    $F445:   LDA #$0C
    $F447:   ADC $C4
    $F449:   TAY
    $F44A:   BNE -$24 ; -> $F428
    $F428:   LDA $FC18,Y
    $F42B:   STA $9E,X
    $F42D:   INY
    $F42E:   INY
    $F42F:   INY
    $F430:   INY
    $F431:   LDA $FC18,Y
    $F434:   STA $98,X
    $F436:   INY
    $F437:   INY
    $F438:   INY
    $F439:   INY
    $F43A:   LDA $FC18,Y
    $F43D:   STA $AC,X
    $F43F:   INX
    $F440:   CPX #$02
    $F442:   BEQ $08 ; -> $F44C
    $F44C:   RTS
    $F0D5:   LDX #$01
    $F0D7:   STX $0A
    $F0D9:   INX
    $F0DA:   STX $A7
    $F0DC:   JMP $F550
    $F550:   LDX #$0A
    $F552:   LDY #$00
    $F554:   LDA $E5
    $F556:   BPL $02 ; -> $F55A
    $F55A:   JSR $F560
    $F560:   LDA $00A8,Y
    $F563:   AND #$F0
    $F565:   LSR
    $F566:   ADC $FC32
    $F569:   STA $AE,X
    $F56B:   LDA #$FC
    $F56D:   STA $AF,X
    $F56F:   DEX
    $F570:   DEX
    $F571:   LDA $00A8,Y
    $F574:   AND #$0F
    $F576:   ASL
    $F577:   ASL
    $F578:   ASL
    $F579:   ADC $FC32
    $F57C:   STA $AE,X
    $F57E:   LDA #$FC
    $F580:   STA $AF,X
    $F582:   INY
    $F583:   DEX
    $F584:   DEX
    $F585:   BPL -$27 ; -> $F560
    $F560:   LDA $00A8,Y
    $F563:   AND #$F0
    $F565:   LSR
    $F566:   ADC $FC32
    $F569:   STA $AE,X
    $F56B:   LDA #$FC
    $F56D:   STA $AF,X
    $F56F:   DEX
    $F570:   DEX
    $F571:   LDA $00A8,Y
    $F574:   AND #$0F
    $F576:   ASL
    $F577:   ASL
    $F578:   ASL
    $F579:   ADC $FC32
    $F57C:   STA $AE,X
    $F57E:   LDA #$FC
    $F580:   STA $AF,X
    $F582:   INY
    $F583:   DEX
    $F584:   DEX
    $F585:   BPL -$27 ; -> $F560
    $F560:   LDA $00A8,Y
    $F563:   AND #$F0
    $F565:   LSR
    $F566:   ADC $FC32
    $F569:   STA $AE,X
    $F56B:   LDA #$FC
    $F56D:   STA $AF,X
    $F56F:   DEX
    $F570:   DEX
    $F571:   LDA $00A8,Y
    $F574:   AND #$0F
    $F576:   ASL
    $F577:   ASL
    $F578:   ASL
    $F579:   ADC $FC32
    $F57C:   STA $AE,X
    $F57E:   LDA #$FC
    $F580:   STA $AF,X
    $F582:   INY
    $F583:   DEX
    $F584:   DEX
    $F585:   BPL -$27 ; -> $F560
    $F587:   LDX #$08
    $F589:   LDY $FC3C
    $F58C:   LDA $B0,X
    $F58E:   CMP $FC32
    $F591:   BNE $06 ; -> $F599
    $F593:   STY $B0,X
    $F595:   DEX
    $F596:   DEX
    $F597:   BPL -$0D ; -> $F58C
    $F58C:   LDA $B0,X
    $F58E:   CMP $FC32
    $F591:   BNE $06 ; -> $F599
    $F593:   STY $B0,X
    $F595:   DEX
    $F596:   DEX
    $F597:   BPL -$0D ; -> $F58C
    $F58C:   LDA $B0,X
    $F58E:   CMP $FC32
    $F591:   BNE $06 ; -> $F599
    $F593:   STY $B0,X
    $F595:   DEX
    $F596:   DEX
    $F597:   BPL -$0D ; -> $F58C
    $F58C:   LDA $B0,X
    $F58E:   CMP $FC32
    $F591:   BNE $06 ; -> $F599
    $F593:   STY $B0,X
    $F595:   DEX
    $F596:   DEX
    $F597:   BPL -$0D ; -> $F58C
    $F58C:   LDA $B0,X
    $F58E:   CMP $FC32
    $F591:   BNE $06 ; -> $F599
    $F593:   STY $B0,X
    $F595:   DEX
    $F596:   DEX
    $F597:   BPL -$0D ; -> $F58C
    $F599:   RTS
    $F55D:   JMP $F59A
    $F59A:   LDA #$FF
    $F59C:   STA $2C
(B) $F59E:   LDX $0284
```

**Comment by thrust26** [comment]
Acid Drop gets stuck in a timer loop, which includes WSYNC. So the timer is checked every 76 cycles. I an odd number is read, it can never read 0. Panda Chase is similar, here the loop is 12 cycles long.

Probably the timer is off by just 1 cycle.

**Comment by thrust26** [comment]
In Stella 4.7.3, the timer switches back to 64T after a (large) number of time underflows. That's why the games work.

**Comment by sa666666** [comment]
Yes, that's probably the effect of the "!(timer & 0x40000)" condition.  But I'm thinking this isn't correct behaviour, which is why the code was changed to "timer >= 0".

**Comment by thrust26** [comment]
There is a noticeable delay after pressing fire in Panda Chase before the picture is displayed again. Does this happen on real hardware too?

**Comment by DirtyHairy** [comment]
> Does this happen on real hardware too?

It doesn't feel like it, the transition seems instant to me on real hardware, while the delay on Stella is evident. The switch back to 64T is definitiely an effect of the old condition (`"!(timer & 0x40000)`) and the reason why those ROMs work in old Stella after all. I, too, think that this is not correct behavior, but more of an "overflow with benefits" :smirk:

In Mr. Roboto, the decisive change is the RSYNC implementation that affects the phase between CPU and TIA and, via WSYNC, in turn affects the phase between CPU and timer. As for Panda Chase, there is no RSYNC and no WSYNC involved.

**Comment by sa666666** [comment]
There are several variations of 'Panda Chase' that also exhibit the same issue.  Not surprising, really, since they're basically the same ROM with perhaps only colours and/or logo changed.

**Comment by thrust26** [comment]
I wrote a simple test program to understand what's really going on on real hardware. But as of now the results are more than confusing: After the (64T) timer has run down to 0, it continues with to count down every 64 cycles. This is obviously **completely** different to the documented and well tested behavior. Since the test code works perfectly in Stella, there must be a weird bug in my test code that I yet have to hunt down and fix. Maybe I can identify some undocumented function that way.

**Comment by thrust26** [comment]
Now I am officially lost. 

Attached you find the result of my test program and the code itself (the timer check logic is in GameCalc subroutine). In the middle column of the output you can see the delta timer value at the top and the following 40 timer values (in intervals of 13 cycles). The left and right playfields show the end result of the loop counters (running from 255 to 0). 
The timer is initialized with 1 and then the code waits three scan lines to be sure there is an underflow. So in theory the timer should run in intervals of 1 cycle when the test code starts.

At the top you see the output of 4.7.3. The loop ends before the loop counters reach zero and the timer counts in intervals of 64 cycles (decreasing every 4-5 values). This is as expected.
![testtimer_4_7_3](https://cloud.githubusercontent.com/assets/24360774/23808990/2a7bd9fc-05cc-11e7-8e01-29ebd67352ff.png)

Below the final output of 5.0 pre5. The loop ends when loop counters reach zero and the timer counts in intervals of 1 cycles (decreasing by 13 every value). This is as expected too.
![testtimer_5_0_pre5](https://cloud.githubusercontent.com/assets/24360774/23808985/27d7ab86-05cc-11e7-83ed-20ccba1ca25c.png)

The last picture shows the (faked by setting the initial timer to $ad instead of 1) output on my TV. The loop ends immediately, with a delta of 13 cycles and the timer counts in intervals of 64 cycles. This would mean, that the timer does **NOT** count in 1 cycle intervals after reaching zero. **And that's where I am lost.** Any ideas???
![testtimer_2600](https://cloud.githubusercontent.com/assets/24360774/23809528/48b84f3e-05ce-11e7-9c65-ab2d340d2903.png)

BTW: Adding up to 256 lines of WSYNC before checking the timer doesn't change the general behavior at all. The timer still counts in 64 cycle intervals.

[TimerTest_001.zip](https://github.com/stella-emu/stella/files/834922/TimerTest_001.zip)

**Comment by sa666666** [comment]
Should I be seeing the same output as you for these test runs?  Because I'm not getting the same output.

**Comment by thrust26** [comment]
Yes.

What do you get?

**Comment by sa666666** [comment]
First screen from real console (light-sixer), also the same on all versions of Stella:
![1](https://cloud.githubusercontent.com/assets/16061066/23812080/8cabca06-05b3-11e7-9e13-467fb2ee0b3b.png)

Second screen from real console (light-sixer):
![2](https://cloud.githubusercontent.com/assets/16061066/23812085/8cb5955e-05b3-11e7-9fb7-fe582a9fa179.png)

Second screen from Stella 4.7.3:
![2-473](https://cloud.githubusercontent.com/assets/16061066/23812081/8cac0368-05b3-11e7-9d6b-d638da709889.png)

Second screen from latest git (5.0.0-pre5 + last few commits):
![2-latest_git](https://cloud.githubusercontent.com/assets/16061066/23812084/8caf3e48-05b3-11e7-8920-65bdb46c0cd7.png)

So at least compared to my system, neither Stella 4.7.3 nor the latest code is completely correct.

**Comment by sa666666** [comment]
Output from Stellarator is closest to the real thing (but the middle column has some pixels rearranged).

Output from z26 matches Stella 4.7.3.

Output from MESS matches Stella git.

I don't know what to say about this :disappointed:

**Comment by thrust26** [comment]
Yea, there is a delay of about 2 seconds (don't ask me why) before the test is done. So ignore the first screen, please.  I should have mentioned that, sorry.

For the second screens, there seems to be a small difference in the results (the code I posted seems not to be 100% identical with the one I used for my tests, but the general behavior is identical.

The question is: Why does the timer on the real console count in steps of 64 cycles?

**Comment by sa666666** [comment]
From Stellarator:
![screenshot_20170310_173441](https://cloud.githubusercontent.com/assets/16061066/23813104/e6a23cbc-05b7-11e7-8717-04d58d626c2c.png)

**Comment by sa666666** [comment]
From Javatari:
![screenshot_20170310_173718](https://cloud.githubusercontent.com/assets/16061066/23813183/3df18d60-05b8-11e7-9a89-2b16ecb4c976.png)

**Comment by thrust26** [comment]
Bingo! That's like the real console!

**Comment by sa666666** [comment]
Javatari matches the real console the closest, with only the ends of the last 4 lines in the middle column differing from my console.

I can't answer the question of why it's happening, but at least we can look at the source for Stellerator and Javatari to see what they do differently.

**Comment by sa666666** [comment]
Acid Drop and Panda Chase also work correctly (and don't lock after pressing Fire) in Javatari, but still lock in Stellerator.  So output-wise, Stellerator is very close, but Javatari is almost exact in both output and the ROMs actually working.

**Comment by thrust26** [comment]
The initial read value on the real console and in Javatari are both $E6. But the internal counter in the real console is ~32 cycles higher, so the 64 cycle timer is decreased 2-3 rows later.

**Comment by thrust26** [comment]
The Javatari in code Pia.js looks pretty straightforward as far as I can tell. And the comment says:
_// If timer underflows, return to 1 cycle interval per specification_

But somehow it emulates differently.

**Comment by ale-79** [comment]
Look at the attached "KIM HINTS" doc for the Kim-1. The Kim-1 uses a 6530 ROM-RAM-I/O-TIMER (RRIOT), and I believe that the timer part works the same as the 6532 in the 2600. According to this document (page 4), if you read the timer from the "TIM64T" addresses ($296 and $29E in the case of the  2600), **the divide ratio is restored to its previously programmed value**, the interrupt option is disabled and the timer keeps its current count.


[kim1_hints.pdf](https://github.com/stella-emu/stella/files/835175/kim1_hints.pdf)

**Comment by thrust26** [comment]
Good catch! That would be this code from Javatari:

```    
    var readFromINTIM = function() {
        INSTAT &= 0xbf;	// Resets bit 6 (Overflow since last INTIM read)
        // If fastDecrement was active (currentTimerInterval == 1), interval always returns to set value after read per specification
        if (currentTimerInterval === 1)
            timerCount = currentTimerInterval = lastSetTimerInterval;
    };
```    
It starts making sense...

**Comment by DirtyHairy** [comment]
Seems I am late to the party :smirk: This is a very interesting find, and it is definitely not part of the datasheets I have been looking at.

**Comment by sa666666** [comment]
I've never heard of this before either.  I used both the Rockwell R6532 and Commodore NMOS 6532 datasheets, and neither mention anything like this.  Hmm, always something new to learn about the 2600 :smiley:

**Comment by DirtyHairy** [comment]
I have made an experiment and implemented this behavior into 6502.ts; check out the current build of stellerator. Though not quite there yet, the improvement is obvious, though not quite there yet. Unsurprisingly, Panda Chase and Acid Drop work with this change :smirk:

![fix1](https://cloud.githubusercontent.com/assets/2444224/23816344/93e5ef66-05ec-11e7-8ac1-3de95f1b144e.png)

The surplus lines from Javatari are there, and there is an additional gap in the first line, but there are five blocks in the lowest right column (as on real hardware, and opposed to Javatari). I have to go to bed now, but I'll try to revisit this later this weekend.

**Comment by thrust26** [comment]
Don't focus on the bottom value to much. The top value is slightly too high ($E7 instead of $E6). And values below are all slightly too low. That's why the bottom row is $9F instead of $A0..

BTW: Are you from Germany too?

**Comment by DirtyHairy** [comment]
OK, that makes sense. I'll guess I'll be taking a look at the source to understand a bit better what I am actually seeing here later 😊

Indeed I am -  from Würzburg, to be precise 😉

**Comment by thrust26** [comment]
After sorting out the problem, I simplified my test program. Now it uses TIM8T and simply reads the timer every 7 cycles. That way we can exactly identify if the internal counter is correct too. 

Attached the (faked) output of my console and the test code with ROM.
![testtimer_2600_8](https://cloud.githubusercontent.com/assets/24360774/23821866/811122f4-063f-11e7-9407-012bc4ec14a6.png)
The internal timer of the latest Stellerator is slightly off. Its 4 cycles too low, maybe that is related to the timer divide ratio...?
![stellarator](https://cloud.githubusercontent.com/assets/24360774/23821891/23010728-0640-11e7-9bbf-16d131a1b839.png)

[TimerTest_002.zip](https://github.com/stella-emu/stella/files/835624/TimerTest_002.zip)

**Comment by thrust26** [comment]
The same as above, but now using TIM64T (and more samples).

Real 2600:
![testtimer_2600_64](https://cloud.githubusercontent.com/assets/24360774/23821991/642ec2d8-0642-11e7-8ed9-a49f9feb0c5b.png)
Here the internal timer of Stellerator ~48 cycles too low:
![stellarator64](https://cloud.githubusercontent.com/assets/24360774/23821993/69eeb2dc-0642-11e7-947b-da238bc04735.png)
[TimerTest_003.zip](https://github.com/stella-emu/stella/files/835635/TimerTest_003.zip)

**Comment by thrust26** [comment]
BTW: I went through the 6532 documentation Stephen has mentioned above. With the new knowledge, you can find the information hidden there:
_"After the interrupt flag is set the internal clock begins counting down at the system clock rate..."._ 
And a few paragraphs later it says:
 _"The interrupt flag will be reset whenever the Timer is accessed by a read or write."_

😊

**Comment by sa666666** [comment]
Hmm, there it is.  I guess hindsight really is 20/20; it was there all along :smiley:

**Comment by DirtyHairy** [comment]
Woha, I have read those two lines, but I never would have deducted _that_ :smile:

Actually, thanks to your tests, Thomas, I think I have nailed it. The current stellerator build seems to perfectly match all three tests to me (including my Jr.). I guess what is really going on in that chip is that, after wrapping and switching to increment-by-one, the configured divider still gets clocked and, after INTIM is read, resumes driving the counter in whatever state it is this point.

As the tests measure a beating pattern, it is hard to tell for sure the fix is true, but I am verry confident that this actually is correct (the way to be sure would be to let the counter wrap, read INTIM and then measure the precise number of cycles it needs to wrap again). If we all agree, I'll port this behavior to Stella in the next few days.

**Comment by thrust26** [comment]
I wrote a dead simple program to test what you describe in your last paragraph. I tested with 1T, 8T, 64T 1024T and for all Stellerator works 100% like my console.

So your theory is spot on. And I think Javatari would work 100% correct too if it wouldn't assign the lastSetTimerInterval to timerCount.

Case closed! :)

**Comment by DirtyHairy** [comment]
05b10be reworks the RIOT implementation to match real hardware. @thrust26 , could you attach your test programs? The implementation was tricky, and it would be great to make sure that there is no +-1 bug lurking somwhere :smirk:

@sa666666 could you take a second look at the commit? The changes affect the debugger, and I am not completely sure that I have preserved the semantics of the "total clocks" and "INTIM clocks" watches.

**Comment by thrust26** [comment]
Attached
[TimerTest_004.zip](https://github.com/stella-emu/stella/files/840513/TimerTest_004.zip)

**Comment by sa666666** [comment]
I won't be able to look at this until this evening, or possibly even tomorrow morning.  However, on first glance it concerns me that some of the const accessor methods are changed to non-const.  The problem is that the accessor methods for intim() and timint() can be called multiple times from the debugger (ie, when a frame is redrawn, etc), but they should not change the internal state of the RIOT class.  Removing the const seems to imply that state **is** changed.

But like I said, I haven't had a chance to properly review this yet ...

**Comment by DirtyHairy** [comment]
@sa666666 I removed the `const` because they **do** change state --- they fast-forward the RIOT state to reflect the current emulation time in terms of system clocks :smirk: Before my change, the internal state of the RIOT that relates to those accessors did not change with time (the change was calculated on the fly and discarded).

However, the methods are still idempotent and do not execute any side effects that would be triggered by readong from the bus, so calling them multiple times from the debugger isn't going to cause any undesired side effects.

**Comment by sa666666** [comment]
OK, I did find some time to look at this after all.  Everything looks fine, and the read-only methods are still read-only, even if not marked as const.  I've done all the testing I can, and everything is working fine.

I updated the bracing in the new code to match what was in the rest of the M6532 class (sorry, just OCD kicking in I guess), and also updated the changelog.  So unless anyone else has issues, I think this can be closed.
